Go/Go.sublime-syntax

```diff
@@ -967,175 +1059,269 @@ contexts:
+  #
+  # Note: this should not be confused with `pop-type-parameter-list` that
+  # uses "typed" syntax.
+  pop-type-parameter-list-simple:
+    - match: \[
+      scope: punctuation.section.brackets.begin.go
+      set: pop-type-parameter-list-simple-inner
+    - include: pop-before-nonblank
+
+  pop-type-parameter-list-simple-inner:
+    - include: pop-on-bracket-end
+    - include: match-identifier-blank
+    - match: '{{ident}}'
+      scope: variable.parameter.type.go
+    - include: match-any
+
+  # Note: this should not be confused with `pop-type-parameter-list-simple` or
+  # `pop-type-argument-list`.
+  pop-type-parameter-list:
+    - match: \[
+      scope: punctuation.section.brackets.begin.go
+      set: pop-type-parameter-list-inner
+    - include: pop-before-nonblank
+
+  # Unlike value parameter lists that have both named and anonymous modes,
+  # type parameter lists have only the named mode. Parameter names must exist.
+  pop-type-parameter-list-inner:
+    - include: pop-on-bracket-end
+    - match: '{{ident_blank}}(?={{noise}},)'
+      scope: variable.language.blank.go
+    - match: '{{ident}}(?={{noise}},)'
+      scope: variable.parameter.type.go
+    - match: '{{ident_blank}}'
+      scope: variable.language.blank.go
+      push: pop-type
+    - match: '{{ident}}'
+      scope: variable.parameter.type.go
+      push: pop-type
+    - include: match-any
+
+  # Note: this should not be confused with `pop-type-parameter-list`.
+  pop-type-argument-list:
+    - include: pop-before-terminator
+    - match: \[
+      scope: punctuation.section.brackets.begin.go
+      set: pop-type-argument-list-inner
+    - include: pop-before-nonblank
+
+  pop-type-argument-list-inner:
+    - include: pop-on-bracket-end
+    - include: match-identifier-blank
+    - match: '{{ident}}'
+      scope: variable.other.type.go
+    - include: match-any
+
+  pop-now:
+    - match: ''
+      pop: true
+
   pop-before-nonblank:
     - match: (?=\S)
       pop: true
 
-  pop-on-semicolon:
-    - match: ;
-      scope: punctuation.terminator.go
+  pop-before-semicolon:
+    - match: (?=;)
+      pop: true
+
+  pop-before-newline:
+    - match: (?=$)
+      pop: true
+
+  pop-before-terminator:
+    - include: pop-before-semicolon
+    - include: pop-before-newline
+
+  pop-before-paren-end:
+    - match: (?=\))
+      pop: true
+
+  pop-on-paren-end:
+    - match: \)
+      scope: punctuation.section.parens.end.go
+      pop: true
+
+  pop-before-bracket-end:
+    - match: (?=\])
+      pop: true
+
+  pop-on-bracket-end:
+    - match: \]
+      scope: punctuation.section.brackets.end.go
+      pop: true
+
+  pop-on-brace-end:
+    - match: \}
+      scope: punctuation.section.braces.end.go
       pop: true
 
-  pop-on-terminator:
-    - include: pop-on-semicolon
-    - match: $
+  pop-before-brace-end:
+    - match: (?=\})
       pop: true
 
+  match-type:
+    - match: (?=\S)
+      push: pop-type
```

Comment:
I'd suggest the following to remove one extra context switch. I think this trivial kind of duplication doesn't hurt.

```suggestion
    - match: (?=\S)
      push: [pop-after-type, pop-type-single]
```